### DISCUSSION TITLE
Next auth integration with existing Node/Express/JWT Backend.

### DISCUSSION THREAD

**Discussion by caioricciuti** [nextauthjs/next-auth#7876] [Help] [answered]
Hello all, hope all is good. I didn't find any straight answer to my question[ (Here is the closest) ](https://github.com/nextauthjs/next-auth/discussions/1290 )so here it goes: 

My backend runs on `Node.js`  and `Express` I have implemented authentication using `JWT`. It's pretty straightforward I check the credentials over my `Mysql` db and if all is good I send back an `access token` and one `refresh token`. 

My question is how can I use Next auth with this approach?

 I've been trying for the last couple of days but I didn't succeed... 

That's what I have til now: 
```js
export const authOptions = {
  providers: [
    Credentials({
      name: "Credentials",
      credentials: {
        email: { label: "email", type: "email" },
        password: { label: "Password", type: "password" },
      },
      authorize: async (credentials) => {
        const res = await fetch(`${process.env.API_URL}/auth`, {
          method: "POST",
          body: JSON.stringify(credentials),
          headers: { "Content-Type": "application/json" },
        });
        const auth = await res.json();
        const user = auth.user;
        if (res.ok && user.jwt) {
          return {
            jwt: user.jwt,
            refresh: user.refreshToken
            name: user.name,
            email: user.email,
          };
        } else {
          return null;
        }
      },
    }),
  ],
  pages: {
    signIn: "/signin",
  },
};

const handler = NextAuth(authOptions);

export { handler as GET, handler as POST };
```

With this, code I can log in with a user, but I get this as session: 
```json
{ "user": { "name": "Example User", "email": "user@example.com" }, "expires": "2023-07-25T12:34:11.710Z" }
```
I don't get the `jwt` not the `refresh token` so I can add both when sending requests to my backend

Is there anyone that had the same problem? Is this the best approach? Should I simply try to create my own provider and not use Next auth??

Thank you all in advance!

**Comment by galaxxyz** [comment, marked as answer]
Hi! The **user** field of your **session** will only contain `name`, `email` and `image` by default. They are populated from the return object of the `authorize(...)`. 

If you need to expose any other data in session, you have to specify it in [callbacks](https://next-auth.js.org/configuration/callbacks). In my case I simply add the whole `user` object  I receive from my backend (which contain **jwt**) to `session.user`.

```typescript
export const options: NextAuthOptions = {
  providers: [
    // ... your providers options
  ],
  callbacks: {
    async jwt({ token, user, account, profile }) {
      return { ...token, ...user }
    },
    async session({ session, token }) {
      session.user = token as any
      return session
    },
  },
}
```

This way I can get jwt from sesson on the server side:
```typescript
const session = await getServerSession(options)
// --> session.user.jwt
```

Be careful, because it is possible to access session on the client as well, using **useSession** hook. By calling this hook you expose your session data (tokens and other potentially sensitive information) to the client side.
```typescript
const session = useSession()
``` 

Related [question](https://github.com/nextauthjs/next-auth/issues/7976).

**Reply by caioricciuti** [reply]
@galaxxyz thank you very much for your response! I'll try this approach, also good to know the security breaches you mentioned. Thanks again!
